Implement RFC 3139: alternative registry authentication support

arlosi opened the pull request:
Allows registries to request Cargo to send the authentication token for all requests, rather than just publish/yank, implementing [RFC 3139](https://github.com/rust-lang/cargo/issues/10474).

### Items from the [tracking issue](https://github.com/rust-lang/cargo/issues/10474)

> Do registries need a more fine-grained switch for which API commands require authentication?

This PR uses the `auth_required` boolean as described in the RFC.

> The RFC mentions adding --token to additional commands like install and search

These flags are not added by this PR.

> Consider changing the name and form of the X- header

Changed to the `www-authenticate` header as suggested by the comments.

> Will there be any concerns with the interaction with https://github.com/rust-lang/rfcs/pull/3231

Not that I know of.

-------------

Adds a new field `"auth-required": true` to `config.json` that indicates Cargo should include the token in all requests to a registry.

For HTTP registries, Cargo first attempts an un-authenticated request, then if that fails with HTTP 401, an authenticated request is attempted. The registry server may include a `www-authenticate` header with the HTTP 401 to instruct Cargo with URL the user can visit to acquire a token (crates.io/me).

Since the API URL is not known (because it's stored in the index), the unstable credential provider feature is modified to key off the index url, and the registry name is no longer provided.

To handle the case where an alternative registry's name is not known (such as coming from a lock file, or via `--index`), Cargo can now look up the token in the configuration by matching on the index URL. This introduces a new error if two alternative registries are configured with the same index URL.

Several operations, such as `cargo install` could have had a `--token` argument added, however it appears that Cargo would like to move away from passing the token on the command line for security reasons. In this case, users would need to configure the registry via the config file (or environment variables) when using `cargo install --index ...` or similar.

rust-highfive commented:
r? @ehuss

(rust-highfive has picked a reviewer for you, use r? to override)

arlosi commented on src/doc/src/reference/unstable.md:
```diff
@@ -1066,8 +1066,8 @@ interactions are:
 * `CARGO` — Path to the `cargo` binary executing the command.
-* `CARGO_REGISTRY_NAME` — Name of the registry the authentication token is for.
-* `CARGO_REGISTRY_API_URL` — The URL of the registry API.
+* `CARGO_REGISTRY_NAME_OPT` — Optional name of the registry. Should not be used as a storage key. Not always available.
```
It may be better to just remove `CARGO_REGISTRY_NAME_OPT`, thoughts?

arlosi replied:
I've removed it.

Eh2406 replied:
I thought Eric asked you to keep it. https://github.com/rust-lang/cargo/pull/10592#issuecomment-1135113946

arlosi commented on src/cargo/util/auth.rs:
```diff
@@ -0,0 +1,444 @@
+    Ok(Some(token))
+}
+
+enum Action {
```
From this point down in the file, it's based on existing code moved from `ops/registry/auth.rs`

arlosi commented on src/cargo/sources/registry/mod.rs:
```diff
@@ -543,12 +552,10 @@ impl<'cfg> RegistrySource<'cfg> {
+        assert!(source_id.is_remote_registry());
         let name = short_name(source_id);
         let ops = if source_id.url().scheme().starts_with("sparse+") {
-            if !config.cli_unstable().http_registry {
```
This unstable check was moved into `HttpRegistry::new`

arlosi commented on src/cargo/sources/registry/http_remote.rs:
```diff
@@ -344,40 +425,57 @@ impl<'cfg> RegistryData for HttpRegistry<'cfg> {
-        }
-
-        if self.config.offline() {
-            return Poll::Ready(Err(anyhow::anyhow!(
```
This check was unnecessary, since it's already done by `ops::http_handle` below.

arlosi commented on src/cargo/util/mod.rs:
```diff
@@ -107,3 +108,15 @@ pub fn indented_lines(text: &str) -> String {
         .collect()
 }
+
+pub fn truncate_with_ellipsis(s: &str, max_length: usize) -> String {
```
This was moved from `registry/ops.rs`

Eh2406 commented on tests/testsuite/registry.rs:
```diff
@@ -1117,26 +1116,10 @@ fn login_with_token_on_stdin() {
+    cargo_process("login")
+        .with_stdout("please paste the token found on [..]/me below")
+        .with_stdin("some token")
+        .run();
```
This is much cleaner. Thank you.

Eh2406 commented on src/cargo/sources/registry/http_remote.rs:
```diff
@@ -631,6 +706,116 @@ impl<'cfg> Downloads<'cfg> {
+
+    /// Parse a `www-authenticate` header into a map of "challenges" and their parameters.
+    /// See IETF RFC 7235 https://datatracker.ietf.org/doc/html/rfc7235#section-4.1 for details.
+    pub fn parse<T: AsRef<str>>(header_values: &[T]) -> HashMap<String, HashMap<String, String>> {
```
I do not have a lot of confidence that this bespoke parser works correctly in all cases. Maybe it doesn't matter given that were only using it for a error message URL? (Although I would like to use it for the "challenge" in https://github.com/rust-lang/cargo/issues/10519) Is there some way to simplify it because we only care about data related to the `Cargo` name? Or use one of the parsing tools Cargo already depends on? Or is there a test suite of real `www-authenticate` headers we could test it on?

Eh2406 replied:
Also [`<auth-scheme>`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#directives) is case insensitive. So we need to handle that somewhere.

Eh2406 replied:
Maybe use https://github.com/scottlamb/http-auth/blob/main/Cargo.toml with default=false, need to ask the team.

arlosi replied:
Okay, I've switched to use the `http-auth` crate.

Eh2406 commented on tests/testsuite/registry_auth.rs:
```diff
@@ -0,0 +1,247 @@
+    let p = make_project();
+
+    // Two alternative registries with the same index.
+    cargo(&p, "build")
```
I think there's room to have a discussion here about whether this error should occur when the configuration is constructed, or when an ambiguous configuration is read.
- The advantage to construction time (Like this PR imps), is that it is clear to the user what is causing the problem.
- The advantage of read time, is that in most contexts this is not ambiguous. It only becomes ambiguous if you have multiple configured registries on the same URL and a registry depends on something else from that URL.

We can always loosen it in the future. So this is probably the right call. But I do want to bring up the issue.

arlosi replied:
We could also be able to attempt to disambiguate by excluding registries that don't have tokens configured.

As you mentioned, I did it this way because it seemed easier to explain to the user.

Eh2406 commented on tests/testsuite/alt_registry.rs:
```diff
@@ -706,11 +707,6 @@ fn no_api() {
-    p.cargo("login --registry alternative TOKEN")
-        .with_status(101)
-        .with_stderr_contains(&err)
-        .run();
```
Why is this removed?

Eh2406 commented on tests/testsuite/build.rs:
```diff
@@ -2864,33 +2864,6 @@ fn recompile_space_in_name() {
-    perms.set_mode(0o000);
-    fs::set_permissions(credentials, perms).unwrap();
-
-    p.cargo("build").run();
```
Why is this removed?

arlosi replied:
With this change, `cargo` always reads the `credentials` file when it loads the other configuration.

I'm not sure what the intent behind supporting an unreadable credentials file is?

Eh2406 replied:
I honestly don't remember either, and I suspect we're comfortable making this "breaking" change.
I will try and remember to bring it up at a Cargo Meeting, next time there's any room on the agenda.

Eh2406 replied:
For reference PR #7774 and Issue #7624

arlosi commented on tests/testsuite/alt_registry.rs:
```diff
@@ -706,11 +707,6 @@ fn no_api() {
-    p.cargo("login --registry alternative TOKEN")
-        .with_status(101)
-        .with_stderr_contains(&err)
-        .run();
```
I should be able to put it back. Thanks for asking.

`cargo login` fetches `config.json` from the registry index in order to determine the API url so it can construct the `/me` URL. That failure could have been caused by missing token. Now that we have the `AuthorizationError` error type, I think I can specifically detect that case, and allow this to become a failure again.

Eh2406 commented on src/cargo/sources/registry/http_remote.rs:
```diff
@@ -111,8 +124,8 @@ struct Download {
-    /// ETag or Last-Modified header received from the server (if any).
-    index_version: RefCell<Option<String>>,
+    /// HTTP headers.
+    header_map: RefCell<HashMap<String, Vec<String>>>,
```
Should we have a HashMap here, or list out the headers we actually care about? Either a variable per header we care about, or a data structure with each of them as separate fields.

arlosi commented:
@Eh2406 I updated this PR to address your comments. I've also cleaned up the test framework to combine the two test http servers (index and api) into one, and added additional tests.

Eh2406 commented:
This is looking good. I want to collect the questions, and talk it over at the next available Cargo meeting. (Hopefully tomorrow.)

Thanks for all the hard work!

Eh2406 commented:
We were able to discuss this PR briefly at yesterday's meeting. I am trying to recall what we were able to discuss and what we weren't, and try and assign people to follow up.

We did not particularly talk about:

> Since the API URL is not known (because it's stored in the index), the unstable credential provider feature is modified to key off the index url, and the registry name is no longer provided.

> It may be better to just remove `CARGO_REGISTRY_NAME_OPT`

Some more context: This is an environment variable that is provided to a credential provider when a registry name is known. Given that it's never technically necessary, and not consistently available, should we just remove it?

> This introduces a new error if two alternative registries are configured with the same index URL.

Some more context: The RFC authorized us to make this an error. This PR implements the error eagerly. This has the advantage that the moment you corrupt your configuration you will get an error message. And the implementation is pretty straightforward, load all configuration and look for conflicts. It has the disadvantage that ms-configuring registry `foo` will prevent you from building a project that depends only on crates.io.

@rust-lang/cargo any thoughts on the above questions?

We were able to briefly discuss:

> With this change, cargo always reads the credentials file when it loads the other configuration.

Some more context: This is a consequence of the eager evaluation above. We need to load all registry configurations in order to figure out if any of them conflict, to fully load a registry we need to look at credentials to see if there's a token associated with it.

In the meeting: We would like to not read credentials if we don't have to, but are not 100% committed to it. we were wondering what the alternative designs would be here, and what their disadvantages would be? Could we eagerly load config's but lazily load credentials? @arlosi what are your thoughts?

> New dep on https://github.com/scottlamb/http-auth with default=false

Some more context: Cargo needs some way for a server to tell it how a user can get a token. It would be nice to follow the `www-authenticate` standard, but it is a bit of a mess. The PR originally had 200 lines of parsing code, but I was not confident in this being completely correct. We found a dependency that does this parsing and switch to using that.

In the meeting: We would like to do a little more research on alternatives to this dependency. One alternative is to parse ourselves using one of the libraries we already use. (@epage you got the most experience with parsing, does this make any sense?) Alternatively we would like to shop around to see how this crate compares to other crates for this purpose. Specifically, What crates are there? And for each one:
- How actively maintained as it / does the code look up to snuff?
- Is it widely used? Reverse dependencies / download counts.
- Is the API plausible for our use?
- Is the license reasonable for our use?
- Are the dependencies reasonable?

We may decide that the one I picked originally is correct, but someone should document the due diligence. @arlosi what are your thoughts, would you be up to collecting this data? (If not I can redo the work, and this time document it.)

arlosi commented:
### Could we eagerly load config's but lazily load credentials?

Cargo currently works by [eagerly loading credentials](https://github.com/rust-lang/cargo/blob/a4c1cd0eb6b18082a7e693f5a665548fe1534be4/src/bin/cargo/commands/publish.rs#L31) for operations that are known to require them (publish, etc). The credentials are then merged into the `Config`.

This change makes it so the credentials *may* be required for pretty much anything hitting the network. If we wanted to continue to lazy-load the credentials, we'd probably want to do it at the point someone called [`auth::registry_credential_config`](https://github.com/arlosi/cargo/blob/bbadf83da0def8ceab6fc1ad596c40aaa90b131e/src/cargo/util/auth.rs#L19).

Unfortunately, the `Config` is not mutable, which makes it difficult to merge additional values into it at that point.

### Should we use the `http-auth` crate to parse the `www-authenticate` header?
* License is OK (Apache 2, or MIT)
* Reverse dependencies are minimal (only 3 small crate)
* Download count is small (3k)
* Code seems actively maintained and good quality
* Has no new transitive dependencies (with default-features = false)
* We're only using a subset of crate functionality. It can do other things such as respond to `Basic` and `Digest` auth (if those features are enabled).

If we're going to use an existing external crate, this is definitely the best one.

The alternative is the [parser I wrote](https://github.com/rust-lang/cargo/blob/be5caeeb7facef633d751c6b202e37591aeeafd9/src/cargo/sources/registry/http_remote.rs#L723) within cargo.
* Less well tested
* Avoids a dependency
* Makes best effort to parse
* 69 lines of code (excluding tests)

This PR currently uses the `http-auth` crate.

Eh2406 commented on src/cargo/util/auth.rs:
```diff
@@ -0,0 +1,427 @@
+        let credential = if process.is_some() && config.cli_unstable().credential_process {
+            let process = process.unwrap();
+            RegistryCredentialConfig::Process((process.path.resolve_program(config), process.args))
+        } else if let Some(token) = config.get_string("registry.token")?.map(|p| p.val) {
```
The old code has a blocker for the case when crates.io has both a credential process and a token. https://github.com/rust-lang/cargo/blob/0f75da631b4284180232c5a50bdf37e714acd059/src/cargo/ops/registry.rs#L404 I am not seeing the corresponding check here.
If this has slipped through, we probably need a test for this case.

arlosi replied:
Hmm. `registry.credential-process`, is used as both as the default credential process for alternative registries, and the credential process for crates.io.

I wonder if we should separate these? For now, I think we just need to add the error (and a test) back as you're suggesting here. But it's something to think about.

Eh2406 replied:
In the process of re-basing my asymmetric token work on yours I isolated my fix for this problem,
https://github.com/rust-lang/cargo/commit/1406b798d9b680b2e0d1dc3df6472c8a6f208589
If that code is useful use it. If not ignore it.

arlosi replied:
Thanks! I've taken your fix to the issue.

ehuss commented on src/cargo/sources/registry/http_remote.rs:
```diff
@@ -269,6 +297,50 @@ impl<'cfg> HttpRegistry<'cfg> {
+                Poll::Ready(Err(anyhow::anyhow!("config.json not found in registry")))
+            }
+            Poll::Ready(LoadResponse::CacheValid) => {
+                panic!("config.json is not stored in the index cache")
```
Can this use `cargo::util::internal(…)` instead? That way, the user gets a nicer message asking them to file a bug report.

ehuss commented on src/cargo/sources/registry/http_remote.rs:
```diff
@@ -402,19 +518,33 @@ impl<'cfg> RegistryData for HttpRegistry<'cfg> {
 
-        // Make sure we don't send data back if it's the same as we have in the index.
         let mut headers = List::new();
+        headers.append("cargo-protocol: version=1")?;
```
I haven't seen any discussion about what this `cargo-protocol` header is. Can you say more about it?

arlosi replied:
It was a suggestion in the RFC process. It would enable the server to potentially do something different if we ever change how the protocol works. It currently doesn't do anything.

ehuss replied:
Can you add a comment to that effect to explain what it is for?

ehuss commented on src/cargo/sources/registry/http_remote.rs:
```diff
@@ -642,7 +739,6 @@ mod tls {
         if ptr == 0 {
             f(None)
         } else {
-            // Safety: * `ptr` is only set by `set` below which ensures the type is correct.
```
Any particular reason this comment is being removed?

ehuss reviewed (commented):
A mostly cursory review of just a few things I noticed.

ehuss commented:
Thanks so much for working on this, it looks great! I particularly like the updates to the tests.

Some notes:

- I think it would be good to keep the `CARGO_REGISTRY_NAME_OPT` option, and to use that to store as an informational element in the key store. That way, if a user looks at the key store they can see something that has a little more meaning than just a URL.
- It looks like the `1password` credential program is broken. It looks like the search code isn't working correctly, as it is looking at the title, but this was changed to remove the title. I think it needs some kind of title. I'm not sure if the search code can search by URL. Can you make sure that all of the credential programs continue to work?
- `-Z registry_auth` needs to be documented in `unstable.md`.
- Can you make sure to update the version of crates-io to `0.35.0`, and `cargo-credential` to `0.2.0`?
- I would be much more comfortable if credentials were loaded only as needed. I imagine there are several ways that could be done, but one idea is to have two `values` tables (the normal `Config.values` and a new one for credentials). The `get_cv` function could then consult the credentials one first, and then the normal one. In the common case where the credentials aren't loaded, I don't think it should impact performance noticeably. I'm not sure I fully follow what `Config.credential_cache` is currently doing, but perhaps it could use that?

arlosi commented on src/cargo/core/source/source_id.rs:
```diff
@@ -251,6 +251,11 @@ impl SourceId {
     }
 
+    /// Gets the name of the registry as defined in the config.
+    pub fn cfg_name(&self) -> Option<&str> {
```
In what way is it unreliable? If it's set, then I expected it to be valid. I can add a comment urging caution that it's not always available.

It's used for printing the help message suggesting how to log in `cargo login --registry <name>`, and internally in the `login` operation to know what the registry's name is when saving.

arlosi commented:
Thanks for all the feedback. I'll get started on figuring out how to lazy load the `credentials` file.

For 1password:
* The `op` tool requires a paid 1password account, which I don't have.
* They have recently released a `v2` of the `op` tool. Do we want to switch to that? I can't find documentation for the v1 protocol. https://developer.1password.com/docs/cli/reference/management-commands/item

I can test the other credential providers.

ehuss commented:
Hm, I didn't know there was a new version of the 1password client. It would probably be good to update it, but that can probably be done separately.

The key issue is that there needs to be a way to uniquely find the credential. The `search` function currently filters based on the title which was the registry name. I think perhaps that can be changed to search for the url, something like:

```diff
diff --git a/crates/credential/cargo-credential-1password/src/main.rs b/crates/credential/cargo-credential-1password/src/main.rs
index 86fc9fed8..bd713fd0e 100644
--- a/crates/credential/cargo-credential-1password/src/main.rs
+++ b/crates/credential/cargo-credential-1password/src/main.rs
@@ -41,7 +41,7 @@ struct ListItem {

 #[derive(Deserialize)]
 struct Overview {
-    title: String,
+    url: String,
 }

 impl OnePasswordKeychain {
@@ -175,11 +175,7 @@ impl OnePasswordKeychain {
         Ok(buffer)
     }

-    fn search(
-        &self,
-        session: &Option<String>,
-        registry_name: &str,
-    ) -> Result<Option<String>, Error> {
+    fn search(&self, session: &Option<String>, index_url: &str) -> Result<Option<String>, Error> {
         let cmd = self.make_cmd(
             session,
             &[
@@ -196,15 +192,15 @@ impl OnePasswordKeychain {
             .map_err(|e| format!("failed to deserialize JSON from 1password list: {}", e))?;
         let mut matches = items
             .into_iter()
-            .filter(|item| item.overview.title == registry_name);
+            .filter(|item| item.overview.url == index_url);
         match matches.next() {
             Some(login) => {
                 // Should this maybe just sort on `updatedAt` and return the newest one?
                 if matches.next().is_some() {
                     return Err(format!(
-                        "too many 1password logins match registry name {}, \
+                        "too many 1password logins match registry URL {}, \
                         consider deleting the excess entries",
-                        registry_name
+                        index_url
                     )
                     .into());
                 }
@@ -232,6 +228,8 @@ impl OnePasswordKeychain {
                 "Login",
                 &format!("password={}", token),
                 &format!("url={}", index_url),
+                "--title",
+                "Cargo registry token",
                 "--tags",
                 CARGO_TAG,
             ],
```

arlosi commented:
@ehuss

> I think it would be good to keep the CARGO_REGISTRY_NAME_OPT option

I've added it back, and used it in the credential process crates

> It looks like the 1password credential program is broken

I implemented your suggestion to fix it, but I don't have a 1password account to test it. I've manually verified the `wincred` and `gnome-secret` providers.

> -Z registry_auth needs to be documented in unstable.md.

Done.

> Can you make sure to update the version of crates-io to 0.35.0, and cargo-credential to 0.2.0?

Yes, good idea. I updated the versions of all the credential providers as well.

> I would be much more comfortable if credentials were loaded only as needed.

I implemented this as you suggested by having an additional `LazyCell` on the `Config` that allows the credentials to be loaded only when needed and re-added the test.

bors commented:
:umbrella: The latest upstream changes (presumably #10738) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Eh2406 commented on crates/cargo-test-support/src/registry.rs:
```diff
@@ -516,10 +536,6 @@ impl HttpServer {
-            if method == "stop" {
-                // Shutdown the server.
-                return;
-            }
```
Did you mean to remove this?

bors commented:
:umbrella: The latest upstream changes (presumably #10753) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #10764) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Eh2406 commented:
@ehuss I'd appreciate your review when you have a chance. It should be easier now that the test code has already been merged.

arlosi commented on src/cargo/util/config/mod.rs:
```diff
@@ -174,6 +177,9 @@ pub struct Config {
     updated_sources: LazyCell<RefCell<HashSet<SourceId>>>,
+    /// Cache of credentials from configuration or credential providers.
+    /// Maps from url to credential value.
+    credential_cache: LazyCell<RefCell<HashMap<CanonicalUrl, String>>>,
```
Since this now can potentially make 100s of requests to authenticated sparse registries, we need somewhere to cache the credential. This is especially true with credential providers, since we don't want to call it for every request.

AngusWaller commented:
It seems like this PR is falling behind. My organization requires this functionality for its registry and I have been trying to sell them on Rust for a while. 

Is there anything I can do to help this move along?

arlosi commented:
It's not abandoned. I've been working on writing and implementing [RFC 3289](https://github.com/rust-lang/rfcs/pull/3289) to clean up how source replacement works around authentication. Will be getting back to this soon.

arlosi commented:
I updated this PR to be based on top of #10907 as two commits.

AngusWaller commented:
Oh awesome news! I hope it gets in soon

bors commented:
:umbrella: The latest upstream changes (presumably #10977) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Eh2406 commented:
I noticed when experimenting with this that sometimes outgoing requests redact the authenticate header, and other requests do not. We should probably consistently redact the header.

arlosi commented:
Do you mean the `Authorization` header? What do you mean by "redact the header"? Are you talking about log messages?

Eh2406 commented:
Sorry. I should have spent more time drafting that message.
Yes I was talking about the `Authorization` header in log messages.
I will get more details and edit them in this message.

bors commented:
:umbrella: The latest upstream changes (presumably #11068) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #11205) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

arlosi commented on src/cargo/sources/registry/http_remote.rs:
```diff
@@ -354,40 +433,68 @@ impl<'cfg> RegistryData for HttpRegistry<'cfg> {
+                        for challenge in http_auth::ChallengeParser::new(header) {
+                            match challenge {
+                                Ok(challenge) if challenge.scheme.eq_ignore_ascii_case("Cargo") => {
+                                    // Look for the `login_url` parameter.
```
I believe this functionality already works. The only place that `/me` is hard coded is `cargo login`, which already handles the special `AuthorizationError` error type and pulls out the `login_url`

bors commented:
:umbrella: The latest upstream changes (presumably #11307) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #11285) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

weihanglo commented on src/cargo/ops/registry.rs:
```diff
@@ -807,22 +740,36 @@ fn http_proxy_exists(config: &Config) -> CargoResult<bool> {
+    let reg_cfg = auth::registry_credential_config(config, &source_ids.original)?;
+    let login_url = match registry(config, token, None, reg, false, false) {
+        Ok((registry, _)) => Some(format!("{}/me", registry.host())),
+        Err(e) if e.is::<AuthorizationError>() => e
```
I am not aware of any. There was a request for making error more structural. https://github.com/rust-lang/cargo/issues/10160

bors commented:
:umbrella: The latest upstream changes (presumably #11369) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

arlosi commented on src/cargo/util/auth.rs:
```diff
@@ -0,0 +1,477 @@
+            })
+            .filter_map(|(k, v)| Some((k, CanonicalUrl::new(&v.into_url().ok()?).ok()?)))
+            .filter(|(_, v)| v == index)
+            .map(|(k, _)| k.to_lowercase())
```
I added a block comment here to attempt to clarify things further about why it's implemented this way.

Given the potential for user confusion in the unlikely case where environment variables and config values are potentially getting mixed, I'll upgrade the `debug` log to be printed by default if the name of the registry in the environment variable doesn't match. That should make it easier to debug if there's an issue.

```
            config.shell().note(format!(
                "name of alternative registry `{}` set to `{name}`",
                sid.url()
            ))?
```

ehuss commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 9827412fee4f5a88ac85e013edd954b2b63f399b has been approved by `ehuss`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=ehuss 9827412fee4f5a88ac85e013edd954b2b63f399b -->
<!-- homu: {"type":"Approved","sha":"9827412fee4f5a88ac85e013edd954b2b63f399b","approver":"ehuss","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit 9827412fee4f5a88ac85e013edd954b2b63f399b with merge 4d5c036a005073794ea880103409afd04ff61649...
<!-- homu: {"type":"BuildStarted","head_sha":"9827412fee4f5a88ac85e013edd954b2b63f399b","merge_sha":"4d5c036a005073794ea880103409afd04ff61649"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/3484551952/jobs/5829754741)
Approved by: ehuss
Pushing 4d5c036a005073794ea880103409afd04ff61649 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"ehuss","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/3484551952/jobs/5829754741"},"merge_sha":"4d5c036a005073794ea880103409afd04ff61649"} -->

jonathanstrong commented:
very happy to see this merged! awesome work @arlosi and @ehuss! this change is a critical improvement enabling private registries like Shipyard.rs.

fasterthanlime commented:
Can the `-Z registry-auth` option be passed to cargo as an environment variable? Right now (with cargo 2022-11-21 + shipyard.rs), my rust-analyzer setup is broken because... it invokes `cargo metadata` without `-Z registry-auth` (that's one of the rust-analyzer cargo invocations you can't override):

![image](https://user-images.githubusercontent.com/7998310/203077490-89ec9b2a-593a-4282-96c2-6223ed3e5a5e.png)

Eh2406 commented:
Yes, the env is `CARGO_UNSTABLE_REGISTRY_AUTH=true`

jonathanstrong commented:
just noting another option, in `~/.cargo/config.toml`:

```toml
[unstable]
registry-auth = true
```

fasterthanlime commented:
@Eh2406 @jonathanstrong are either of these options documented somewhere? I've reached the "reading cargo sources" stage of banging my head against the wall here, maybe https://doc.rust-lang.org/cargo/reference/environment-variables.html could point somewhere useful?

jonathanstrong commented:
> please be mindful that the current state of nightly is likely to change. The plan is for the stabilize version of the feature to require the use of asymmetric tokens for authenticated registries.

is there any place this decision is being discussed/debated? would like to weigh in and I imagine this is not the best venue!

arlosi commented:
> is there any place this decision is being discussed/debated?

Feel free to start a thread on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo) if you'd like to chat, or create a new issue if there's something specific you'd like changed.

You can also keep adding comments here too.

ehuss commented on tests/testsuite/registry_auth.rs:
```diff
@@ -0,0 +1,304 @@
+  unable to get packages from source
+
+Caused by:
+  multiple registries are configured with the same index url \
```
@dasbard It's better to file a new issue instead of commenting no a closed PR.  I opened #11524 to track it.
